doc/statemachines.tex

```diff
@@ -0,0 +1,310 @@
+\item[Upstream,Downstream,Topology]
+\end{description}
+
+\begin{description}
+\item[State Machine]
+  The Protocols are described as a state machine.
+  Formally the state machine is a tuple $(S,T,initial,final)...$ (States, Transitions, initial state, finial state)...
+  The specification uses different representations of the state machine that all describe the same protocol
+  but may leave out some aspects.
+  \begin{itemize}
+  \item The state machine can be represented as a collection of tables (States, Transitions/Messages..).
+  There is a one to one correspondence between these tables and the Haskell source.
+\item A transition graph of the global view of the state machine.
+\item A transition graph of the client view.
+\item A transition graph of the server view.
+\end{itemize}
+  
+\item[States]
+  The specification assumes a global view of the state of a system consisting of the client and the server.
+  Client and server are always in identical states (by definition)
+  which also means that client and server simultaneously transit to new states.
+  The global state is not observable by the client nor the server.
+  Possible interpretations: 
+  A sequence of states is a set of points in time that a correct run of the protocol passes through.
+  (todo: draw diagram)
+\item[Messages/Events]
+  A tuple $(lable,data)$. Can be serialized transferred over the network and de-serialized.
+  (authenticity of messages not handled here)
+\item[Transitions]
+  The document describes the global/abstract state transitions of the protocol
+  in terms of messages and data that the client and the server exchange.
+  In sync with state transitions the  client and server process the data locally to compute the
+  the actual "outcome" of a protocol-run. (For example "update the block chain").
+  Parts of the local computations are specific to the implementation language and the underlying
+  data structures, but they are also relevant for the protocol (to determine the next transition).
+\item[Initial state]
+\item[Final state \& termination]
+\end{description}
+  
+\subsubsection{Protocol Description}
+
+
+\section{The Chain Synchronization Protocol}
+This section describes the Chain Synchronization Protocol.
+
+\begin{description}
+\item[purpose of the protocol]
+The Chain Synchronization Protocol is used by the block chain consumer
+to synchronize its block chain with the block chain producer.
+\item[Who is communicating]
+  A node communicates with several upstream and downstream node.
+  The node runs an independent agents for every other node
+  it communicates with. The typically a node runs agents where it acts as a server and also instances
+  where it acts as a client.
+\item[Where is located in the protocol stack ?]
+\item[What are the interfaces to the lower layers of the stack ?]
+  The chain synchronization protocol can send and receive messages. (via a point to point channel.)
```

Comment:
I think it would be good, with time (or place dummy sections) to extend each section describing a protocol to include serialization of it's messages.  By the way, we already have [cbor codec](https://github.com/input-output-hk/ouroboros-network/blob/master/ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/Codec/Cbor.hs) for chain following protocol, and I am just working on the [block download codec](https://github.com/input-output-hk/ouroboros-network/tree/coot/block-fetch-codec/ouroboros-network/src/Ouroboros/Network/Protocol/BlockFetch)